plugins/shorewall/shorewall_log

```diff
@@ -0,0 +1,292 @@
+
+    return 'yes'
+
+
+if len(sys.argv) == 2 and sys.argv[1] == "autoconf":
+    print(autoconf())
+    sys.exit()
+
+
+from collections import defaultdict, namedtuple
+from typing import Set, Iterator, TextIO, Dict, Tuple
+from itertools import takewhile
+from subprocess import run, PIPE
+import shlex
+import shutil
+import pickle
+import re
+
+
+logfile = os.getenv('logfile', 'journald' if shutil.which('journalctl') else '/var/log/kern.log')
+journalctl_args = list(shlex.split(os.getenv('journalctlargs',
+                                             'SYSLOG_IDENTIFIER=kernel')))
+taggroups = os.getenv('taggroups')
+taggroups = taggroups.split() if taggroups else []
+tagfilter = os.getenv('tagfilter')
+tagfilter = tagfilter.split() if tagfilter else []
+include_ungrouped = (os.getenv('includeungrouped', 'true').lower() == 'true')
+prefix_format = os.getenv('prefixformat')
+if prefix_format:
+    if sys.version_info < (3, 7):
+        prefix_format = re.escape(prefix_format).replace('\\%s', '(.+)').replace('\\%d', '\\d+')
+    else:  # % is no longer escaped
+        prefix_format = re.escape(prefix_format).replace('%s', '(.+)').replace('%d', '\\d+')
+    prefix_format = re.compile(prefix_format)
+
+
+def get_logtags() -> Tuple[Dict[str, Set[str]], Dict[str, Set[str]]]:
+    rules = (run(['iptables-save'], stdout=PIPE, universal_newlines=True).
+             stdout.splitlines())
+    tags = defaultdict(set)
+    groups = defaultdict(set)
+
+    for line in rules:
+        args = iter(shlex.split(line))
+        for arg in args:
+            if arg == '-j' and next(args) != 'LOG':
+                break
+            if arg == '--log-prefix':
+                prefix = next(args)
+
+                if prefix_format:
+                    tag = prefix_format.match(prefix)[1]
+                else:
+                    tag = prefix.rstrip()
+
+                for ignored in tagfilter:
+                    if ignored in tag:
+                        continue
+                tags[tag].add(prefix)
+
+                for group in taggroups:
+                    if group in tag:
+                        groups[group].add(tag)
+                        break
+                else:
+                    if include_ungrouped:
+                        groups[tag].add(tag)
+
+                break
+
+    return groups, tags
+
+
+State = namedtuple('State', ['journal', 'file'])
+
+
+def load_state() -> State:
+    try:
+        with open(os.getenv('MUNIN_STATEFILE'), 'rb') as f:
+            return pickle.load(f)
+    except OSError:
+        return State(None, None)
+
+
+def save_state(state: State):
+    with open(os.getenv('MUNIN_STATEFILE'), 'wb') as f:
+        return pickle.dump(state, f)
+
+
+def get_lines_journalctl(state: State) -> Iterator[str]:
+    cursor = state.journal
+
+    def catch_cursor(line: str):
+        cursor_id = '-- cursor: '
+        if line.startswith(cursor_id):
+            save_state(State(line[len(cursor_id):-1], None))
```

Comment:
Is the last character in `line` thrown away on purpose? Maybe it is a line break?

Reply:
that was a bug, `.splitlines()` already removes the line breaks